File: two_pointer/557_reverse_words_in_a_string_iii.py
class Solution:
    def reverseWords(self, s: str) -> str:
        words = list(s)
        l = 0
        for r in range(len(words)):
            if s[r] == ' ' or r == len(s) - 1:
                left = l
                right = r - 1 if r != len(s) - 1 else r
                while left < right:
                    words[left], words[right] = words[right], words[left]
                    left += 1
                    right -= 1
                l = r + 1
        return ''.join(words)



        

File: string/557_reverse_words_in_a_string_iii.java
class Solution {
    public String reverseWords(String s) {
        int left = 0, right = 0, n = s.length();
        char[] res = s.toCharArray();
        while (left < n) {
            while (right < n && res[right] != ' ') {
                right++;
            }
            reverse(res, left, right - 1);
            left = right + 1;
            right = left;
        }
        return new String(res);
    }
    
    public void reverse(char[] res, int start, int end) {
        while (start < end) {
            char temp = res[start];
            res[start] = res[end];
            res[end] = temp;
            start++;
            end--;
        }
    }
}